File: src/routes/images.ts
import { Router, Request, Response } from "express";
import fs from "fs";
import path from "path";

import { prisma } from "../prisma";
import { getLibrary } from "../lib/library";
import { asyncHandler } from "../middleware/asyncHandler";
import { ApiError } from "../types/api";

export const imagesRouter = Router();

imagesRouter.get(
  "/cover-image/:libraryId/:entityId.webp",
  asyncHandler(async (req: Request, res: Response) => {
    const { libraryId, entityId } = req.params;

    if (isNaN(Number(libraryId)) || isNaN(Number(entityId))) {
      throw new ApiError(400, "Invalid library ID or entity ID");
    }

    const library = await getLibrary(libraryId);

    if (!library) {
      throw new ApiError(404, "Library not found");
    }

    if (library.type === "book") {
      const file = await prisma.bookFile.findUnique({
        where: {
          id: Number(entityId),
        },
      });

      if (!file) {
        throw new ApiError(404, "File not found");
      }

      const coverPath = path.join(
        library.path,
        ".devourer",
        "files",
        entityId,
        "cover.webp"
      );

      if (fs.existsSync(coverPath)) {
        const fileBuffer = fs.readFileSync(coverPath);
        res.setHeader("Content-Type", "image/webp");
        res.send(fileBuffer);
      } else {
        throw new ApiError(404, "File not found");
      }
    } else {
      const series = await prisma.mangaSeries.findUnique({
        where: {
          id: Number(entityId),
        },
      });

      if (!series) {
        throw new ApiError(404, "File not found");
      }

      const coverPath = path.join(
        library.path,
        ".devourer",
        "series",
        entityId,
        "cover.webp"
      );

      if (fs.existsSync(coverPath)) {
        const fileBuffer = fs.readFileSync(coverPath);
        res.setHeader("Content-Type", "image/webp");
        res.send(fileBuffer);
      } else {
        throw new ApiError(404, "File not found");
      }
    }
  })
);

imagesRouter.get(
  "/preview-image/:libraryId/:seriesId/:entityId.jpg",
  asyncHandler(async (req: Request, res: Response) => {
    const { libraryId, seriesId, entityId } = req.params;

    if (
      isNaN(Number(libraryId)) ||
      isNaN(Number(seriesId)) ||
      isNaN(Number(entityId))
    ) {
      throw new ApiError(400, "Invalid library ID, series ID or entity ID");
    }

    const library = await getLibrary(libraryId);

    if (!library) {
      throw new ApiError(404, "Library not found");
    }

    const series = await prisma.mangaSeries.findUnique({
      where: {
        id: Number(seriesId),
      },
    });

    if (!series) {
      throw new ApiError(404, "Series not found");
    }

    const file = await prisma.mangaFile.findUnique({
      where: {
        id: Number(entityId),
        series_id: Number(seriesId),
      },
    });

    if (!file) {
      throw new ApiError(404, "File not found");
    }

    const previewPath = path.join(
      library.path,
      ".devourer",
      "series",
      seriesId,
      "previews",
      `${file.file_name}.jpg`
    );

    if (fs.existsSync(previewPath)) {
      const fileBuffer = fs.readFileSync(previewPath);
      res.setHeader("Content-Type", "image/jpeg");
      res.send(fileBuffer);
    } else {
      throw new ApiError(404, "File not found");
    }
  })
);

export default imagesRouter;

File: internal/api/handlers/images.go
package handlers

import (
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/gin-gonic/gin"

	"github.com/devourer/server/internal/db/queries"
)

// CoverImage handles GET /cover-image/:libraryId/:entityId
func (h *Handlers) CoverImage(c *gin.Context) {
	libraryID, err := strconv.ParseInt(c.Param("libraryId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid library ID"})
		return
	}
	entityParam := strings.TrimSuffix(c.Param("entityId"), ".jpg")
	entityID, err := strconv.ParseInt(entityParam, 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid entity ID"})
		return
	}

	lib, err := queries.GetLibraryByID(h.DB, libraryID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Library not found"})
		return
	}

	var coverPath string
	if lib.Type == "book" {
		coverPath = filepath.Join(lib.MetaBase(), "files", strconv.FormatInt(entityID, 10), "cover.jpg")
	} else {
		coverPath = filepath.Join(lib.MetaBase(), "series", strconv.FormatInt(entityID, 10), "cover.jpg")
	}

	if _, err := os.Stat(coverPath); os.IsNotExist(err) {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Cover not found"})
		return
	}
	c.Header("Cache-Control", "public, max-age=3600")
	c.File(coverPath)
}

// PreviewImage handles GET /preview-image/:libraryId/:seriesId/:entityId
func (h *Handlers) PreviewImage(c *gin.Context) {
	libraryID, err := strconv.ParseInt(c.Param("libraryId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid library ID"})
		return
	}
	seriesID, err := strconv.ParseInt(c.Param("seriesId"), 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid series ID"})
		return
	}
	entityParam := strings.TrimSuffix(c.Param("entityId"), ".jpg")
	entityID, err := strconv.ParseInt(entityParam, 10, 64)
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"status": false, "message": "Invalid entity ID"})
		return
	}

	lib, err := queries.GetLibraryByID(h.DB, libraryID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Library not found"})
		return
	}

	file, err := queries.GetMangaFileByID(h.DB, entityID)
	if err != nil || file.SeriesID != seriesID {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "File not found"})
		return
	}

	previewPath := filepath.Join(
		lib.MetaBase(), "series",
		strconv.FormatInt(seriesID, 10),
		"previews",
		file.FileName+".jpg",
	)
	if _, err := os.Stat(previewPath); os.IsNotExist(err) {
		c.JSON(http.StatusNotFound, gin.H{"status": false, "message": "Preview not found"})
		return
	}
	c.Header("Cache-Control", "public, max-age=3600")
	c.File(previewPath)
}
